### PR TITLE
Add a module "leavechans" that allows leaving all the channels matching ...

### PR DESCRIPTION
**Pull request by HeisSpiter**
...specifics names on last client disconnection on all the networks.

### CODE REVIEW

File: modules/leavechans.cpp [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+                {
+                    CIRCNetwork* pNetwork = (*net);
+                    CChan* pChan = pNetwork->FindChan(it->second->GetLeftChan());
+                    if (pChan && pChan->IsOn())
+                    {
+                        PutModule("No one connected to ZNC, leaving " + it->second->GetLeftChan() + ".");
+                        pNetwork->PutIRC("PART " + it->second->GetLeftChan());
+                        if (it->second->GetNotifyMsg().empty())
+                        {
+                            pNetwork->PutIRC("PRIVMSG " + it->second->GetNotifyChan() + " :" + pNetwork->GetCurNick() + " automatically left " + it->second->GetLeftChan() + " due to client disconnection/timeout. Please check current requests.");
```
Comment (DarthGandalf) [inline modules/leavechans.cpp]: Yay for flood!
This is even worse than awaynick

File: modules/leavechans.cpp [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        }
+
+        m_msChans.clear();
+    }
+
+    virtual void OnClientDisconnect()
+    {
+        if (!m_pUser->IsUserAttached())
+        {
+            for (std::map<CString, CLeaveChan*>::const_iterator it = m_msChans.begin(); it != m_msChans.end(); it++)
```
Comment (DarthGandalf) [inline modules/leavechans.cpp]: Why not "auto", or even foreach loop?

File: modules/leavechans.cpp [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+        return true;
+    }
+
+    virtual ~CLeaveChans()
+    {
+        for (std::map<CString, CLeaveChan*>::iterator it = m_msChans.begin(); it != m_msChans.end(); it++)
+        {
+            delete it->second;
+        }
```
Comment (DarthGandalf) [inline modules/leavechans.cpp]: It's easier to store them in the map by value.

File: modules/leavechans.cpp [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    CString m_sLeftChan;
+    CString m_sNotifyChan;
+    CString m_sNotifyMsg;
+};
+
+class CLeaveChans : public CModule {
+public:
+    MODCONSTRUCTOR(CLeaveChans) {}
+
+    virtual bool OnLoad(const CString& sArgs, CString& sMessage)
```
Comment (DarthGandalf) [inline modules/leavechans.cpp]: bool OnLoad(...) override

**Comment by DarthGandalf** [comment]
Why would anyone leave channels on disconnect? I thought bouncers are needed especially to not leave them?

**Comment by teward** [comment]
@HeisSpiter Bouncers are typically designed such that you *don't* leave any channels.  You seem to want to be changing the actual design of what a bouncer does.  What's the actual usable case for this kind of module?

**Comment by HeisSpiter** [comment]
@teward It's been designed for a really specific usage. And as more users have been requesting it, we're proposing it here.

So, basically, it's in use on an IRC network by the IRC staff to make them leave specific channels (such as: support channels) when they are not online. The purpose being to handle IRC staff who would have... poor connection. Which is the purpose of having a ZNC bouncer.
This also explains the notification which is done on line 122: it may happen that the staffer who got disconnected might have an undergoing ticket he was managing. It's here to ask the other staffers to have a look (hence the specific message).

This was the really main reason why it was designed. And finally, some of our users (not staff) started to appreciate it. I don't know their usage, though. But, I believe some other staffs may need it as well?

File: modules/leavechans.cpp (hunk 1 above)
Comment (HeisSpiter) [inline modules/leavechans.cpp]: See usage comment below

File: modules/leavechans.cpp (hunk 2 above)
Comment (HeisSpiter) [inline modules/leavechans.cpp]: This is really old code, before C++11 (it land before ZNC 1.X). A revamp would be doable, for sure.

File: modules/leavechans.cpp (hunk 3 above)
Comment (HeisSpiter) [inline modules/leavechans.cpp]: Noted. If an inclusion is possible, I'll provide a fix there.

File: modules/leavechans.cpp (hunk 4 above)
Comment (HeisSpiter) [inline modules/leavechans.cpp]: Noted. If an inclusion is possible, I'll provide a fix there.

**Comment by teward** [comment]
@HeisSpiter Congrats, you've proven my point.

The module exists for a highly specific use case:  There is one IRC network where the IRC staff need users to leave specific channels when they are not online.  And only some of the users have appreciated the module on that one network.

The use case here is extraordinarily limited, for one network's requirements and needs.

My suggestion is to make this available as a third party module in a separate github repository, taking into account DarthGandalf's suggestions for code above, as well as implementation issues.  The narrow use case is great for one specific network and one specific ZNC instance or only a few, but I don't see any "global" benefit to all ZNC users - only for those on that one network.  With a literally infinite number of networks, that makes the use case for this hyper-narrow and specific, and that is a major negative point against the inclusion of the module, in my opinion.

**Comment by DarthGandalf** [comment]
At least while it autospams channels on every disconnect, inclusion is not possible for sure.

**Comment by teward** [comment]
If I may make a functionality suggestion, at the least, have the module have a delay after client disconnects - this would mean that, say, after 10 seconds of no clients being connected to the bouncer, then part.  This would help protect against join/part spam as a result of bad connections between the user's IRC client and ZNC itself.

(I still don't believe this should be included, but from a module code/actions structure i'll still give some suggestions)

**Comment by Un1matr1x** [comment]
Since this is a realy special purpos, why not make a own github-repo for this module an link it @http://wiki.znc.in/Category:External_Modules - so if anybody is searching for such a module, they can find it quickly

**Pull request closed without merging**